src/realm/index_string.cpp

```diff
@@ -357,46 +358,44 @@ void IndexArray::index_string_all_ins(StringData value, IntegerColumn& result, C
     struct WorkItem {
         const char* header;
         size_t string_offset;
-        int permutation; // -1 == unexpanded, otherwise 0-15
+        key_type key;
     };
     using WorkList = std::vector<WorkItem>;
     WorkList work_list;
 
-    const char* top_header = get_header_from_data(m_data);
-    work_list.push_back({top_header, 0, -1});
-
     const util::Optional<std::string> upper_value = case_map(value, true);
     const util::Optional<std::string> lower_value = case_map(value, false);
 
+    auto generate_search_keys = [&work_list, &upper_value, &lower_value](const char* header,
+                                                                         size_t string_offset) -> void {
+        std::set<key_type> keys_seen;
```

Comment:
I am fully aware of the facts, but admittedly didn't worry about memory allocations etc., because of the life expectancy of the code. That said, I am (also) convinced that a plain,  unsorted `std::vector` is the fastest container for such few elements, operations. I admittedly forgot about `std::unordered_set`. The optimal solution would be to only create combinations of the differing bytes (xor, find 0-bytes etc..), but I found that to be _way too clever_ for something that would go away.